substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrSymbolRepository.java

```diff
@@ -82,22 +99,49 @@ private long getSymbolId(String imageHeapString, boolean replaceDotWithSlash) {
         int hashcode = (int) (rawPointerValue ^ (rawPointerValue >>> 32));
         symbol.setHash(hashcode);
 
-        return table.add(symbol);
+        return getTable().add(symbol);
     }
 
     @Override
     public void write(JfrChunkWriter writer) throws IOException {
-        assert VMOperation.isInProgressAtSafepoint();
+        JfrSymbolHashtable table = getTable(true);
         writer.writeCompressedLong(JfrTypes.Symbol.getId());
         writer.writeCompressedLong(table.getSize());
 
-        // TODO: iterate the table and write the symbol data in the correct encoding. Also consider
-        // symbol.getReplaceDotWithSlash() when writing the data.
+        JfrSymbol[] entries = table.getTable();
+        for (int i = 0; i < entries.length; i++) {
+            JfrSymbol entry = entries[i];
+            if (entry.isNonNull()) {
+                while (entry.isNonNull()) {
+                    writeSymbol(writer, entry);
+                    entry = entry.getNext();
+                }
+            }
+        }
+    }
+
+    private void writeSymbol(JfrChunkWriter writer, JfrSymbol symbol) throws IOException {
+        writer.writeCompressedLong(symbol.getId());
+        writer.writeByte(JfrChunkWriter.TYPE_UTF8);
+        byte[] value = symbol.getValue().getBytes(StandardCharsets.UTF_8);
+        if (symbol.getReplaceDotWithSlash()) {
+            replaceDotWithSlash(value);
+        }
+        writer.writeCompressedInt(value.length);
+        writer.writeBytes(value);
+    }
+
+    private void replaceDotWithSlash(byte[] utf8String) {
+        for (int i = 0; i < utf8String.length; i++) {
+            if (utf8String[i] == '.') {
```

Comment:
This check could result in false positives, i.e., if a character is encoded in multiple bytes, one of the bytes may match. So, when you iterate over the array, you need to skip all characters that are encoded as multiple bytes (`.` is always encoded as a single byte).

Reply:
I was trying to learn about utf8 by reading through https://en.wikipedia.org/wiki/UTF-8#Encoding

For multi-byte characters, the possibilities are 110/111/1111 x... and then 2 to 3 more bytes of 10xxxxxx. All single byte characters are 0xxxxxxx. I think that means there won't be collisions between multi-byte and single-byte characters, right?

Reply:
Indeed, my bad. I wrongly thought that only the very first byte is used to mark multi-byte characters.